src/librustc_typeck/check/coercion.rs

```diff
@@ -424,20 +425,29 @@ impl<'f, 'gcx, 'tcx> Coerce<'f, 'gcx, 'tcx> {
             debug!("Missing Unsize or CoerceUnsized traits");
             return Err(TypeError::Mismatch);
         };
-
-        // Note, we want to avoid unnecessary unsizing. We don't want to coerce to
-        // a DST unless we have to. This currently comes out in the wash since
-        // we can't unify [T] with U. But to properly support DST, we need to allow
-        // that, at which point we will need extra checks on the target here.
+        let mut leftover_predicates = vec![];
 
         // Handle reborrows before selecting `Source: CoerceUnsized<Target>`.
         let (source, reborrow) = match (&source.sty, &target.sty) {
             (&ty::TyRef(_, mt_a), &ty::TyRef(_, mt_b)) => {
                 coerce_mutbls(mt_a.mutbl, mt_b.mutbl)?;
 
-                let coercion = Coercion(self.cause.span);
-                let r_borrow = self.next_region_var(coercion);
-                (mt_a.ty, Some(AutoBorrow::Ref(r_borrow, mt_b.mutbl)))
+                // For `&T`, we don't reborrow, (which is only needed
```

Comment:
Don't you need this also on the raw ptr path? better to subtype `source` as one unit.